### PR TITLE
CIRCSTORE-310 Add holdingsRecordId to request schema

### PR DESCRIPTION
**Pull request by Mykyta-Varenyk**

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/RequestsAPI.java [hunk 1]
```diff
@@ -87,6 +88,10 @@ public void postRequestStorageRequests(
           }
           asyncResultHandler.handle(reply);
         });
+    } else{
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/RequestsAPI.java:91]:
```suggestion
    } else {
```

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java [hunk 2]
```diff
@@ -40,4 +41,32 @@ public static Errors samePositionInQueueError(String item, Integer position) {
 
     return errors;
   }
+
+  public static Errors requestStructureIsValid(Request request){
+    Request.RequestLevel requestLevel = request.getRequestLevel();
+    List<Error> errorList = new ArrayList<>();
+    if (requestLevel == Request.RequestLevel.ITEM) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java:48]: Extract `request.getItemId() == null` and `request.getHoldingsRecordId() == null` into variables

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java [hunk 3]
```diff
@@ -40,4 +41,32 @@ public static Errors samePositionInQueueError(String item, Integer position) {
 
     return errors;
   }
+
+  public static Errors requestStructureIsValid(Request request){
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java:45]:
```suggestion
  public static Errors validateRequest(Request request){
```

File: src/main/java/org/folio/rest/impl/RequestsAPI.java [hunk 4]
```diff
@@ -123,9 +128,11 @@ public void putRequestStorageRequestsByRequestId(
     Map<String, String> okapiHeaders,
     Handler<AsyncResult<Response>> asyncResultHandler,
     Context vertxContext) {
+    Errors errors = RequestsApiUtil.requestStructureIsValid(entity);
 
-    // TODO: On insert don't return 204, we must return 201!
-    MyPgUtil.putUpsert204(REQUEST_TABLE, entity, requestId, okapiHeaders, vertxContext,
+    if(errors.getErrors().isEmpty()) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/RequestsAPI.java:133]:
```
if (errors.getErrors().isEmpty()) {
  asyncResultHandler.handle(succeededFuture(PutRequestStorageRequestsByRequestIdResponse
          .respond422WithApplicationJson(errors)));
  return;
}
```

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java [hunk 5]
```diff
@@ -40,4 +41,32 @@ public static Errors samePositionInQueueError(String item, Integer position) {
...
+    Request.RequestLevel requestLevel = request.getRequestLevel();
+    List<Error> errorList = new ArrayList<>();
+    if (requestLevel == Request.RequestLevel.ITEM) {
+      if (request.getItemId() == null) {
+        Error error = new Error();
+        error.withMessage("ItemId in Item level request should not be null");
+        errorList.add(error);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java:52]: Inline variable

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java [hunk 6]
```diff
@@ -40,4 +41,32 @@ public static Errors samePositionInQueueError(String item, Integer position) {
...
     return errors;
   }
+
+  public static Errors requestStructureIsValid(Request request){
+    Request.RequestLevel requestLevel = request.getRequestLevel();
+    List<Error> errorList = new ArrayList<>();
+    if (requestLevel == Request.RequestLevel.ITEM) {
+      if (request.getItemId() == null) {
+        Error error = new Error();
+        error.withMessage("ItemId in Item level request should not be null");
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java:51]:
```suggestion
        error.withMessage("Item ID in item level request should not be null");
```

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java [hunk 7]
```diff
@@ -40,4 +41,32 @@ public static Errors samePositionInQueueError(String item, Integer position) {
...
+      }
+
+      if (request.getHoldingsRecordId() == null){
+        Error error = new Error();
+        error.withMessage("HoldingsRecordId in Item level request should not be null");
+        errorList.add(error);
+      }
+    } else if (requestLevel == Request.RequestLevel.TITLE &&
+      ((request.getItemId() != null && request.getHoldingsRecordId() == null)
+      || (request.getItemId() == null && request.getHoldingsRecordId() != null))){
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java:62]: Extract method

File: src/test/java/org/folio/rest/api/RequestsApiTest.java [hunk 8]
```diff
@@ -216,6 +218,154 @@ public void canCreateARequest()
     assertThat(tagsRepresentation.getJsonArray("tagList"), contains("new", "important"));
   }
 
+  @Test
+  @Parameters({"Title", "Item"})
+  public void canCreateTitleAndItemLevelRequestsWithHoldingsRecordIdAndItemId(String requestLevel)
+    throws InterruptedException,
+    MalformedURLException,
+    TimeoutException,
+    ExecutionException {
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/api/RequestsApiTest.java:227]: inline

File: src/test/java/org/folio/rest/api/RequestsApiTest.java [hunk 9]
```diff
@@ -926,6 +1076,92 @@ public void cannotUpdateRequestForSameItemToAnExistingPosition()
...
+
+    assertThat(response, hasErrorWith(hasMessageContaining("ItemId in Item level request should not be null")));
+    assertThat(response, hasErrorWith(hasMessageContaining("HoldingsRecordId in Item level request should not be null")));
+  }
+
+  @Test
+  @Parameters(
+    {"holdingsRecordId", "itemId"}
+  )
+  public void cannotCreateTitleLevelRequestIfOneOfItemIdAndHoldingsRecordIdIsNotPresent(String property)
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/api/RequestsApiTest.java:1107]:
```suggestion
  public void cannotCreateTitleLevelRequestIfOneOfItemIdAndHoldingsRecordIdIsNotPresent(String propertyToRemove)
```

File: src/test/java/org/folio/rest/api/RequestsApiTest.java [hunk 10]
```diff
@@ -926,6 +1076,92 @@ public void cannotUpdateRequestForSameItemToAnExistingPosition()
...
+    request.remove(property);
+
+    client.post(requestStorageUrl(), request, TENANT_ID, ResponseHandler.json(createCompleted));
+
+    JsonObject response = createCompleted.get(5, TimeUnit.SECONDS).getJson();
+
+    assertThat(response, hasErrorWith(
+      hasMessageContaining(
+        "In Title level request, there should be either both itemId and holdingsRecordId, " +
+          "or no such fields at all")));
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/api/RequestsApiTest.java:1131]:
```suggestion
        "Title level request must have both itemId and holdingsRecordId or neither")));
```

File: src/test/java/org/folio/rest/support/builders/RequestRequestBuilder.java [hunk 11]
```diff
@@ -227,7 +232,66 @@ public RequestRequestBuilder withRequestType(String requestType) {
       this.position,
       this.pickupServicePointId,
       this.tags,
-      this.patronComments);
+      this.patronComments,
+      this.holdingsRecordId);
+  }
+
+  public RequestRequestBuilder withInstanceId(UUID instanceId) {
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/support/builders/RequestRequestBuilder.java:239]: Replace with Lombok

File: ramls/request.json [hunk 12]
```diff
@@ -225,10 +230,10 @@
   },
   "additionalProperties": false,
   "required": [
-    "itemId",
```
Comment (OleksandrVidinieiev) [inline ramls/request.json:228]: - Revert this change
- add holdingsRecordId (required)

File: src/main/java/org/folio/rest/impl/RequestsAPI.java [hunk 13]
```diff
@@ -76,8 +76,12 @@ public void postRequestStorageRequests(
     Map<String, String> okapiHeaders,
     Handler<AsyncResult<Response>> asyncResultHandler,
     Context vertxContext) {
-
-    PgUtil.post(REQUEST_TABLE, entity, okapiHeaders, vertxContext,
+    Errors errors = RequestsApiUtil.validateRequest(entity);
+    if (!errors.getErrors().isEmpty()) {
+      asyncResultHandler.handle(succeededFuture(PostRequestStorageRequestsResponse
+        .respond422WithApplicationJson(errors)));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/RequestsAPI.java:82]: Add return and move main logic outside of `else`

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java [hunk 14]
```diff
@@ -26,18 +27,46 @@ public static boolean hasSamePositionConstraintViolated(String errorMessage) {
...
     List<Error> errorList = new ArrayList<>();
     errorList.add(error);
 
-    Errors errors = new Errors();
-    errors.setErrors(errorList);
+    return new Errors().withErrors(errorList);
+  }
+
+  public static Errors validateRequest(Request request){
+    Request.RequestLevel requestLevel = request.getRequestLevel();
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java:42]: Static import

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java [hunk 15]
```diff
@@ -26,18 +27,46 @@ public static boolean hasSamePositionConstraintViolated(String errorMessage) {
...
+        errorList.add(createError(
+          "Title level request must have both itemId and holdingsRecordId or neither"));
+      }
+
+    return new Errors().withErrors(errorList);
+  }
+
+  public static boolean hasInvalidFieldCombination(
+    Request request, boolean holdingsRecordIdIsNotPresent, boolean itemIdIsNotPresent){
+    return ((request.getItemId() != null && holdingsRecordIdIsNotPresent)
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java:65]: use arguments with `!`
remove redundant braces

File: src/test/java/org/folio/rest/api/RequestsApiTest.java [hunk 16]
```diff
@@ -926,6 +931,86 @@ public void cannotUpdateRequestForSameItemToAnExistingPosition()
       "Cannot have more than one request with the same position in the queue")));
   }
 
+  @Test
+  public void cannotCreateItemLevelRequestIfItemIdAndHoldingsRecordIdAreNull()
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/rest/api/RequestsApiTest.java:935]: Combine into one awesome test:
- values: valida value, null value, absent
- for both fields

File: src/main/java/org/folio/rest/impl/RequestsAPI.java (hunk 1 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/RequestsAPI.java]: Resolved

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java (hunk 2 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java]: Resolved

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java (hunk 3 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java]: Resolved

File: src/main/java/org/folio/rest/impl/RequestsAPI.java (hunk 4 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/RequestsAPI.java]: Resolved

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java (hunk 5 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java]: Resolved

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java (hunk 7 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java]: Resolved

File: src/test/java/org/folio/rest/api/RequestsApiTest.java (hunk 8 above)
Comment (Mykyta-Varenyk) [reply src/test/java/org/folio/rest/api/RequestsApiTest.java]: Resolved

File: src/test/java/org/folio/rest/api/RequestsApiTest.java (hunk 9 above)
Comment (Mykyta-Varenyk) [reply src/test/java/org/folio/rest/api/RequestsApiTest.java]: Resolved

File: src/test/java/org/folio/rest/api/RequestsApiTest.java (hunk 10 above)
Comment (Mykyta-Varenyk) [reply src/test/java/org/folio/rest/api/RequestsApiTest.java]: Resolved

File: src/test/java/org/folio/rest/support/builders/RequestRequestBuilder.java (hunk 11 above)
Comment (Mykyta-Varenyk) [reply src/test/java/org/folio/rest/support/builders/RequestRequestBuilder.java]: Resolved

File: ramls/request.json (hunk 12 above)
Comment (Mykyta-Varenyk) [reply ramls/request.json:228]: Resolved

File: src/main/java/org/folio/rest/impl/RequestsAPI.java (hunk 13 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/RequestsAPI.java:82]: Resolved

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java (hunk 14 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java]: Resolved

File: src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java (hunk 15 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/util/RequestsApiUtil.java]: Resolved

File: src/test/java/org/folio/rest/api/RequestsApiTest.java (hunk 16 above)
Comment (Mykyta-Varenyk) [reply src/test/java/org/folio/rest/api/RequestsApiTest.java:936]: Resolved
